shared/reducers/profile.tsx

```diff
@@ -35,111 +36,118 @@ const updateUsername = (state: Types.State) => {
+  },
+  [ProfileGen.updateUsername]: (draftState, action) => {
+    draftState.username = action.payload.username
+    updateUsername(draftState)
+  },
+  [ProfileGen.cleanupUsername]: draftState => {
+    updateUsername(draftState)
+  },
+  [ProfileGen.revokeFinish]: (draftState, action) => {
+    draftState.revokeError = action.payload.error ?? ''
+  },
+  [ProfileGen.submitBlockUser]: draftState => {
+    draftState.blockUserModal = 'waiting'
+  },
+  [ProfileGen.finishBlockUser]: (draftState, action) => {
+    draftState.blockUserModal = action.payload.error ? {error: action.payload.error} : undefined
+  },
+  [ProfileGen.updateProofText]: (draftState, action) => {
+    draftState.proofText = action.payload.proof
+  },
+  [ProfileGen.updateProofStatus]: (draftState, action) => {
+    draftState.proofFound = action.payload.found
+    draftState.proofStatus = action.payload.status
+  },
+  [ProfileGen.updateErrorText]: (draftState, action) => {
+    draftState.errorCode = action.payload.errorCode
+    draftState.errorText = action.payload.errorText
+  },
+  [ProfileGen.updateSigID]: (draftState, action) => {
+    draftState.sigID = action.payload.sigID
+  },
+  [ProfileGen.updatePgpInfo]: (draftState, action) => {
+    const valid1 = Validators.isValidEmail(draftState.pgpEmail1)
+    const valid2 = draftState.pgpEmail2 && Validators.isValidEmail(draftState.pgpEmail2)
+    const valid3 = draftState.pgpEmail3 && Validators.isValidEmail(draftState.pgpEmail3)
+    draftState.pgpErrorEmail1 = !!valid1
+    draftState.pgpErrorEmail2 = !!valid2
+    draftState.pgpErrorEmail3 = !!valid3
+    draftState.pgpErrorText = Validators.isValidName(draftState.pgpFullName) || valid1 || valid2 || valid3
+    draftState.pgpFullName = action.payload.pgpFullName ?? draftState.pgpFullName
+    draftState.pgpEmail1 = action.payload.pgpEmail1 ?? draftState.pgpEmail1
+    draftState.pgpEmail2 = action.payload.pgpEmail2 ?? draftState.pgpEmail2
+    draftState.pgpEmail3 = action.payload.pgpEmail3 ?? draftState.pgpEmail3
+  },
+  [ProfileGen.updatePgpPublicKey]: (draftState, action) => {
+    draftState.pgpPublicKey = action.payload.publicKey
+  },
+  [ProfileGen.updatePromptShouldStoreKeyOnServer]: (draftState, action) => {
+    draftState.promptShouldStoreKeyOnServer = action.payload.promptShouldStoreKeyOnServer
+  },
+  [ProfileGen.addProof]: (draftState, action) => {
+    const {platform} = action.payload
+    const maybeNotGeneric = More.asPlatformsExpandedType(platform)
+    draftState.errorCode = undefined
+    draftState.errorText = ''
+    draftState.platform = maybeNotGeneric ?? undefined
+    draftState.platformGeneric = maybeNotGeneric ? undefined : platform
+    updateUsername(draftState)
+  },
+  [ProfileGen.proofParamsReceived]: (draftState, action) => {
+    const {params} = action.payload
+    draftState.platformGenericParams = {
+      ...params,
+      logoBlack: [...params.logoBlack],
+      logoFull: [...params.logoFull],
```

Comment:
Is it necessary to do these spreads? This seems to assume actions can mutate on their own.

Reply:
spoke at length to @buoyad about this earlier today actually. the core issue is the type is `ReadonlyArray<>` but in immer produce the type is just `Array<>` so you can mutate the draft but TS won't allow you to assign a ReadonlyArray to Array as you can then mutate it which is a violation. After much discussion we landed on just using Array<> as the type instead of doing this kind of thing which is mysterious. i'm doing that and changing this